### ISSUE TITLE
DOC503 picks up exception facilities

### ISSUE THREAD

**Issue by nicholasjng** [jsh9/pydoclint#174]
Hey! We're using pydoclint with pre-commit in one of our projects, and got some linter errors with the latest release.

Those errors have to do with us using an "exception translator" function, which takes an exception coming from an API client, translates it into a builtin Python error, and then raises the translated error. For example, an HTTP403 (forbidden) would become a `PermissionError`.

You can find an example here: https://github.com/aai-institute/lakefs-spec/blob/d271df45654a239bacf482338ea8febe5c029886/src/lakefs_spec/spec.py#L356-L357

pydoclint picks this up and reports it as an undocumented exception type. But, of course, we do not want to document this implementation detail, since the user is never supposed to see API errors.

Following this rationale, in the AST analysis, non-exception types (as like in our case, e.g., a function translating exceptions) should not be reported as errors. What do you think?

Error message:

```
    142: DOC503: Method `LakeFSFileSystem.wrapped_api_call` exceptions in the "Raises" section in the docstring do not match those in the function body Raises values in the docstring: ['OSError']. Raised exceptions in the body: ['translate_lakefs_error'].
```

**Comment by nicholasjng** [comment]
As I typed this, I realized that solving this might be more tricky than anticipated, since any translated exception we do document, e.g. the `PermissionError` example from the issue report, never shows up in the AST of the function (it is hidden inside the exception translator).

One option is inlining the translator into the AST before analysis, but that is probably tedious and error-prone. At this stage, I think there's no sound way of supporting this functionality with pydoclint.

**Comment by jsh9** [comment]
Thanks for reporting this issue!

You are correct that there is no (easy) way for pydoclint to "know" what happens under the hood of `translate_lakefs_error()`.  pydoclint does static code analysis using AST, which is very fast and meanwhile ["dumb" by design](https://jsh9.github.io/pydoclint/notes_for_users.html#2-cases-that-pydoclint-is-not-designed-to-handle).

For pydoclint to "know" what `translate_lakefs_error()` returns, we'd need to introduce another Python stdlib module called [`inspect`](https://docs.python.org/3/library/inspect.html), which is much slower than AST.

A workaround for your use case may be like this:

```python
            except NotFoundException:
                # fall through, retry with `ls` if it's a directory.
                pass
            except ServerException as e:
                MyError = translate_lakefs_error(e, rpath=path)
                raise MyError
```
And then you can add `MyError` into your docstring.

**Comment by jsh9** [comment]
> Oh, and could you add DOC503 to the --skip-checking-raises flag? From the name of the switch, it should also swallow DOC503 violations if specified imo.

Right, that's an omission on my part.  I'll make a code change to address this.

> Could there be an option to not err when more exceptions are listed in the docstring than are found in the AST, kind of like saying "I'm aware that those errors I documented do not appear explicitly in the code, but trust me on this"?

Maybe a ["noqa" comment](https://jsh9.github.io/pydoclint/how_to_ignore.html) on this specific function would work?  But note that "noqa" only works in the flake8 mode, not in the native mode.  (There's very minimal overhead with the flake8 mode though.)

**Comment by Marenz** [comment]
I have a variation of this error in the new version that seems also very wrong of me:

tests/test_script.py:97:1: DOC503 Function parse_and_validate exceptions in the "Raises" section in the docstring do not match those in the function body Raises values in the docstring: ['Exception']. Raised exceptions in the body: ['e'].
```
[..]
    Raises:
        Exception: If data is not valid.
[..]
    except Exception as e:
        print(f"Error: {e}")
        print(f"Line: {test_data.timestamp}, {line}")
        raise e

```

Why would it care what the variable of the exception is called?

**Comment by Marenz** [comment]
I suspect [this](https://github.com/jsh9/pydoclint/pull/168) introduced that bug

**Comment by Marenz** [comment]
This also seems to cause other failures across our projects, e.g. https://github.com/frequenz-floss/frequenz-channels-python/actions/runs/11135326252/job/30945126934

**Comment by jsh9** [comment]
Hi @Marenz , I opened another issue (https://github.com/jsh9/pydoclint/issues/175) for what you reported.

**Comment by jsh9** [comment]
Closing this issue as it's a duplicate of https://github.com/jsh9/pydoclint/issues/175

**Comment by jsh9** [comment]
The fix is published in 0.5.11.
